### PR TITLE
Show ECE buttons on products w/o shipping tax included in price, and improve tests

### LINKED ISSUE #9921: Apple Pay Google Pay buttons not shown on products that don't require shipping and prices already include tax.
### Describe the bug
In #9080 these buttons were disabled due to the possibility of not having billing information available when the payment sheet opens and, therefore, we could potentially be showing an incorrect amount if tax is calculated based on billing address. However, an edge case came up in a ticket where the user _did_ have the tax set to be calculated on the billing address for non-shippable products **but** their prices were already entered inclusive of tax. Meaning that no tax calculations should be happening on checkout anyway. I think it should be safe to show the EPMs in this instance. The relevant logic is [here](https://github.com/Automattic/woocommerce-payments/blob/476c528c41c97c867cda9f0b1d3131b3c32fa505/includes/express-checkout/class-wc-payments-express-checkout-button-helper.php#L418-L433).

### To Reproduce
1. Go to `/wp-admin/admin.php?page=wc-settings&tab=tax` and select the radio button to "Yes, I will enter prices inclusive of tax."
2. Set "Calculate tax based on" to "Customer billing address" and save the settings.
3. Add a digital or non-shippable product to the cart.
4. Visit the checkout and the Apple Pay/Google Pay buttons are not added to the page.

### Actual behavior
Currently, with the described tax settings, the EPM buttons are not added

### Screenshots
Image

### Expected behavior
With the tax settings as described above, accurate totals should be able to be sent to the payment sheet and EPM buttons should be enabled.

### PR DESCRIPTION
**Pull request by timur27**
Fixes https://github.com/Automattic/woocommerce-payments/issues/9921

#### Changes proposed in this Pull Request
This PR adds the new part to the conditional responsible for filtering out the ece button based on whether it's not shippable and tax is calculated from billing address, which is not present. 

```php
'yes' !== get_option( 'woocommerce_prices_include_tax' )
```

we now check for the option selected under WooCommerce -> Settings -> Tax -> Prices entered with tax. If prices entered are inclusive of tax, we will consider that ECE button can be rendered even if does not have billing address.

This PR also refactors some tests in this area, because they were mocking the class that is under test and included tests for two classes under one test suite.

#### Testing instructions

* steps to reproduce from https://github.com/Automattic/woocommerce-payments/issues/9921 are performed but on step 4, buttons are displayed on the page

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

File: tests/unit/express-checkout/test-class-wc-payments-express-checkout-button-handler.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		$this->zone->delete_shipping_method( $this->local_pickup_id );
+
+		$this->assertFalse(
+			$this->system_under_test->filter_cart_needs_shipping_address( true ),
+			'Should not require shipping address for subscription without shipping methods'
+		);
+
+		remove_filter( 'woocommerce_shipping_method_count', '__return_zero' );
+		WC_Subscriptions_Cart::set_cart_contains_subscription( false );
+	}
```
Comment (timur27) [inline tests/unit/express-checkout/test-class-wc-payments-express-checkout-button-handler.php:135]: Those are two tests I took from `test-class-wc-payments-express-checkout-button-helper.php` and put into this file as these tests serve different purposes - they test `WC_Payments_Express_Checkout_Button_Handler`

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9954` or branch name `fix/ece-button-for-price-including-tax` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/ece-button-for-price-including-tax) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 036db15a1ca81bdf7eba87fb4dc57ae83a2d5491
- Build time: 2024-12-19 16:38:01 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/express-checkout/class-wc-payments-express-checkout-button-helper.php [hunk 2]
```diff
@@ -415,7 +415,7 @@ public function should_show_express_checkout_button() {
 			return true;
 		}
 
-		// Non-shipping product and billing is calculated based on shopper billing addres. Excludes Pay for Order page.
+		// Non-shipping product and billing is calculated based on shopper billing address. Excludes Pay for Order page.
```
Comment (mdmoore) [inline includes/express-checkout/class-wc-payments-express-checkout-button-helper.php]:
```suggestion
		// Non-shipping product and tax is calculated based on shopper billing address. Excludes Pay for Order page.
```

File: includes/express-checkout/class-wc-payments-express-checkout-button-helper.php [hunk 3]
```diff
@@ -427,7 +427,9 @@ public function should_show_express_checkout_button() {
 			)
 
 			// ...and billing is calculated based on billing address.
```
Comment (mdmoore) [inline includes/express-checkout/class-wc-payments-express-checkout-button-helper.php]:
```suggestion
			// ...and tax is calculated based on billing address.
```

**Review by mdmoore — APPROVED** [review]
LGTM - The ECE is now displayed on non-shippable products when prices include tax and tax is based on billing address. I left a couple very small comments. Thanks Timur!

**Pull request merged**